Concurrent Dictionary for BinaryFormatters

Anipik opened the pull request:
Fixes #25529

Anipik commented:
@danmosemsft  @ViktorHofer  how to show  benchmarkdotnet Results ?

ViktorHofer commented:
There is a markdown (.md) file in the BenchmarkResults folder in your project

Anipik commented:
@ViktorHofer  there is lot of fluctuation between the results of different runs. how should I deal with that ?

ViktorHofer commented:
I suggest to read the BenchmarkDotNet docs. There isn't a straight answer to that

stephentoub commented on src/System.Runtime.Serialization.Formatters/src/System/Runtime/Serialization/Formatters/Binary/BinaryFormatter.cs:
```diff
@@ -91,18 +92,15 @@ internal void Serialize(Stream serializationStream, object graph, bool check)
+                typeInformation = new TypeInformation(FormatterServices.GetClrTypeFullName(type), assemblyName, hasTypeForwardedFrom);
+                s_typeNameCache.TryAdd(type, typeInformation);
             }
+            return typeInformation;
```
This could be written more succinctly using the GetOrAdd method, e.g.:
```C#
internal static TypeInformation GetTypeInformation(Type type) =>
    s_typeNameCache.GetOrAdd(type, t =>
    {
        string assemblyName = FormatterServices.GetClrAssemblyName(t, out bool hasTypeForwardedFrom);
        return new TypeInformation(FormatterServices.GetClrTypeFullName(t), assemblyName, hasTypeForwardedFrom);
    });
```
In addition to being smaller, this also guarantees that the same instance will be returned for a specific type, even if multiple threads race to initialize it.  That's also true in the locking version.  But in the PR version, it's not true because two threads could each create their own typeInformation and return the one they created even if the call to TryAdd fails.

danmosemsft commented:
@anipik if there is fluctuation try giving benchmark.net a higher iteration count. Also make sure the machine is idle when you measure.

danmosemsft commented:
btw Be sure that you're using release bits not debug.

danmosemsft commented:
Waiting on benchmark data proof.

Anipik commented:
``` ini

BenchmarkDotNet=v0.10.10, OS=Windows 10 Redstone 2 [1703, Creators Update] (10.0.15063.726)
Processor=Intel Xeon CPU E5-1650 v4 3.60GHz, ProcessorCount=12
Frequency=3507501 Hz, Resolution=285.1033 ns, Timer=TSC
.NET Core SDK=2.1.1-preview-007118
  [Host]     : .NET Core ? (Framework 4.6.25917.03), 64bit RyuJIT
  Job-TZMMYK : .NET Core 2.1.0-preview1-25919-02 (Framework 4.6.25917.03), 64bit RyuJIT

MaxRelativeError=0.1  RemoveOutliers=True  Runtime=Core  
Toolchain=Core 2.1.0-preview1-25919-02  InvocationCount=10  LaunchCount=1  
TargetCount=10  UnrollFactor=2  WarmupCount=3  

```
|          Method |     Mean |     Error |    StdDev |   Median |      Min |      Max |    Gen 0 |    Gen 1 |    Gen 2 | Allocated |
|---------------- |---------:|----------:|----------:|---------:|---------:|---------:|---------:|---------:|---------:|----------:|
|   serialization | 3.084 ms | 0.0473 ms | 0.0247 ms | 3.080 ms | 3.059 ms | 3.139 ms | 200.0000 | 100.0000 | 100.0000 |   1.86 MB |
| deserialization | 7.509 ms | 0.0558 ms | 0.0332 ms | 7.506 ms | 7.472 ms | 7.572 ms | 700.0000 | 100.0000 | 100.0000 |      5 MB |

Anipik commented:
Without Concurrent Dictionary

``` ini

BenchmarkDotNet=v0.10.10, OS=Windows 10 Redstone 2 [1703, Creators Update] (10.0.15063.726)
Processor=Intel Xeon CPU E5-1650 v4 3.60GHz, ProcessorCount=12
Frequency=3507501 Hz, Resolution=285.1033 ns, Timer=TSC
.NET Core SDK=2.1.1-preview-007118
  [Host]     : .NET Core ? (Framework 4.6.25917.03), 64bit RyuJIT
  Job-TZMMYK : .NET Core 2.1.0-preview1-25919-02 (Framework 4.6.25917.03), 64bit RyuJIT

MaxRelativeError=0.1  RemoveOutliers=True  Runtime=Core  
Toolchain=Core 2.1.0-preview1-25919-02  InvocationCount=10  LaunchCount=1  
TargetCount=10  UnrollFactor=2  WarmupCount=3  

```
|          Method |     Mean |     Error |    StdDev |   Median |      Min |      Max |    Gen 0 |    Gen 1 |    Gen 2 | Allocated |
|---------------- |---------:|----------:|----------:|---------:|---------:|---------:|---------:|---------:|---------:|----------:|
|   serialization | 3.084 ms | 0.0473 ms | 0.0247 ms | 3.080 ms | 3.059 ms | 3.139 ms | 200.0000 | 100.0000 | 100.0000 |   1.86 MB |
| deserialization | 7.509 ms | 0.0558 ms | 0.0332 ms | 7.506 ms | 7.472 ms | 7.572 ms | 700.0000 | 100.0000 | 100.0000 |      5 MB |

Anipik commented:
with Concurrent Dictionary
``` ini

BenchmarkDotNet=v0.10.10, OS=Windows 10 Redstone 2 [1703, Creators Update] (10.0.15063.726)
Processor=Intel Xeon CPU E5-1650 v4 3.60GHz, ProcessorCount=12
Frequency=3507501 Hz, Resolution=285.1033 ns, Timer=TSC
.NET Core SDK=2.1.1-preview-007118
  [Host]     : .NET Core ? (Framework 4.6.25917.03), 64bit RyuJIT
  Job-FYCRXK : .NET Core 2.1.0-preview1-25919-02 (Framework 4.6.25917.03), 64bit RyuJIT

MaxRelativeError=0.1  RemoveOutliers=True  Runtime=Core  
Toolchain=Core 2.1.0-preview1-25919-02  InvocationCount=10  LaunchCount=1  
TargetCount=10  UnrollFactor=2  WarmupCount=3  

```
|          Method |     Mean |     Error |    StdDev |   Median |      Min |      Max |    Gen 0 |    Gen 1 |    Gen 2 | Allocated |
|---------------- |---------:|----------:|----------:|---------:|---------:|---------:|---------:|---------:|---------:|----------:|
|   serialization | 3.196 ms | 0.1885 ms | 0.1247 ms | 3.168 ms | 3.058 ms | 3.351 ms | 200.0000 | 100.0000 | 100.0000 |   1.86 MB |
| deserialization | 7.473 ms | 0.5598 ms | 0.3703 ms | 7.329 ms | 7.151 ms | 8.078 ms | 700.0000 | 100.0000 | 100.0000 |      5 MB |

Anipik commented:
Deserialization include serialization code and then deserializing

ViktorHofer commented:
With how many threads did you test the scenario? Are you able to share the relevant test code so that we can get a better feeling of the results?

ViktorHofer commented:
Shouldn't the allocation differ between `Dictionary<T,U>` and `ConcurrentDictionary<T,U>`? (haven't looked into the sources)

stephentoub commented:
> Shouldn't the allocation differ

It depends on what the test is doing.

Anipik commented:
Code for the tests and benchmarking 
https://github.com/Anipik/Benchmarking

Anipik commented:
@ViktorHofer its a single thread .. I am looking to increase the threads

danmosemsft commented:
@AndresPabloGomez @msisterna  any guess at how many threads are doing binary serialization at the time you hit the perf problems? 2? 5? 50?

AndresPabloGomez commented:
Hi, our server has 24 cores (2 processors Intel(R) Xeon(R) CPU E5-2667 0 @ 2.90GHz). A normal load is 20 threads, a high load is between 90 and 120 threads.
Thanks a lot.

danmosemsft commented:
OK, well @Anipik probably has 8, but that should be sufficient to establish it's an improvement.

danmosemsft commented:
I was expecting to see numbers for the actual multithreaded scenario. The benchmark above just establishes that single threaded isn't regressed.

ViktorHofer commented:
> OK, well @Anipik probably has 8, but that should be sufficient to establish it's an improvement.

Sorry, I interpreted this as you being fine merging it.

danmosemsft commented:
Yeah I could have been clearer -- I meant would be sufficient if he actually uses threads ...  leave it merged, very likely the change is fine, but I'd like to see the numbers pasted here...

Anipik commented:
My system has 6 cores
With Locks 
4 cores

``` ini

BenchmarkDotNet=v0.10.10, OS=Windows 10 Redstone 2 [1703, Creators Update] (10.0.15063.726)
Processor=Intel Xeon CPU E5-1650 v4 3.60GHz, ProcessorCount=12
Frequency=3507499 Hz, Resolution=285.1034 ns, Timer=TSC
.NET Core SDK=2.1.1-preview-007118
  [Host]     : .NET Core ? (Framework 4.6.25917.03), 64bit RyuJIT DEBUG
  Job-RNTFGN : .NET Core 2.1.0-preview1-25919-02 (Framework 4.6.25917.03), 64bit RyuJIT

MaxRelativeError=0.2  RemoveOutliers=True  Runtime=Core  
Toolchain=Core 2.1.0-preview1-25919-02  InvocationCount=6  LaunchCount=1  
TargetCount=2  UnrollFactor=1  WarmupCount=2  

```
|    Method |    Mean | Error |  StdDev |  Median |     Min |     Max |       Gen 0 |      Gen 1 |    Gen 2 | Allocated |
|---------- |--------:|------:|--------:|--------:|--------:|--------:|------------:|-----------:|---------:|----------:|
| serialize | 3.297 s |    NA | 2.289 s | 3.297 s | 1.678 s | 4.915 s | 295166.6667 | 96416.6667 | 250.0000 |  68.28 KB |

Anipik commented:
6 Cores with Locks
``` ini

BenchmarkDotNet=v0.10.10, OS=Windows 10 Redstone 2 [1703, Creators Update] (10.0.15063.726)
Processor=Intel Xeon CPU E5-1650 v4 3.60GHz, ProcessorCount=12
Frequency=3507499 Hz, Resolution=285.1034 ns, Timer=TSC
.NET Core SDK=2.1.1-preview-007118
  [Host]     : .NET Core ? (Framework 4.6.25917.03), 64bit RyuJIT DEBUG
  Job-ATKZVX : .NET Core 2.1.0-preview1-25919-02 (Framework 4.6.25917.03), 64bit RyuJIT

MaxRelativeError=0.2  RemoveOutliers=True  Runtime=Core  
Toolchain=Core 2.1.0-preview1-25919-02  InvocationCount=6  LaunchCount=1  
TargetCount=2  UnrollFactor=1  WarmupCount=2  

```
|    Method |    Mean | Error |  StdDev |  Median |     Min |     Max |       Gen 0 |      Gen 1 |    Gen 2 | Allocated |
|---------- |--------:|------:|--------:|--------:|--------:|--------:|------------:|-----------:|---------:|----------:|
| serialize | 7.422 s |    NA | 5.630 s | 7.422 s | 3.441 s | 11.40 s | 214000.0000 | 70333.3333 | 333.3333 |  41.14 KB |

Anipik commented:
8 cores with Locks

``` ini

BenchmarkDotNet=v0.10.10, OS=Windows 10 Redstone 2 [1703, Creators Update] (10.0.15063.726)
Processor=Intel Xeon CPU E5-1650 v4 3.60GHz, ProcessorCount=12
Frequency=3507499 Hz, Resolution=285.1034 ns, Timer=TSC
.NET Core SDK=2.1.1-preview-007118
  [Host]     : .NET Core ? (Framework 4.6.25917.03), 64bit RyuJIT DEBUG
  Job-TCTOKJ : .NET Core 2.1.0-preview1-25919-02 (Framework 4.6.25917.03), 64bit RyuJIT

MaxRelativeError=0.2  RemoveOutliers=True  Runtime=Core  
Toolchain=Core 2.1.0-preview1-25919-02  InvocationCount=6  LaunchCount=1  
TargetCount=2  UnrollFactor=1  WarmupCount=2  

```
|    Method |    Mean | Error |  StdDev |  Median |     Min |     Max |       Gen 0 |      Gen 1 |    Gen 2 | Allocated |
|---------- |--------:|------:|--------:|--------:|--------:|--------:|------------:|-----------:|---------:|----------:|
| serialize | 8.556 s |    NA | 5.580 s | 8.556 s | 4.610 s | 12.50 s | 249750.0000 | 94583.3333 | 500.0000 |  55.43 KB |

Anipik commented:
4,6,8 cores respectively with Concurrent Dictionary
``` ini

BenchmarkDotNet=v0.10.10, OS=Windows 10 Redstone 2 [1703, Creators Update] (10.0.15063.726)
Processor=Intel Xeon CPU E5-1650 v4 3.60GHz, ProcessorCount=12
Frequency=3507499 Hz, Resolution=285.1034 ns, Timer=TSC
.NET Core SDK=2.1.1-preview-007118
  [Host]     : .NET Core ? (Framework 4.6.25917.03), 64bit RyuJIT DEBUG
  Job-DGNHBI : .NET Core 2.1.0-preview1-25919-02 (Framework 4.6.25917.03), 64bit RyuJIT

MaxRelativeError=0.2  RemoveOutliers=True  Runtime=Core  
Toolchain=Core 2.1.0-preview1-25919-02  InvocationCount=6  LaunchCount=1  
TargetCount=2  UnrollFactor=1  WarmupCount=2  

```
|    Method |    Mean | Error |  StdDev |  Median |     Min |     Max |       Gen 0 |      Gen 1 |    Gen 2 | Allocated |
|---------- |--------:|------:|--------:|--------:|--------:|--------:|------------:|-----------:|---------:|----------:|
| serialize | 3.920 s |    NA | 3.767 s | 3.920 s | 1.256 s | 6.584 s | 152750.0000 | 50500.0000 | 333.3333 |  28.03 KB |
``` ini

BenchmarkDotNet=v0.10.10, OS=Windows 10 Redstone 2 [1703, Creators Update] (10.0.15063.726)
Processor=Intel Xeon CPU E5-1650 v4 3.60GHz, ProcessorCount=12
Frequency=3507499 Hz, Resolution=285.1034 ns, Timer=TSC
.NET Core SDK=2.1.1-preview-007118
  [Host]     : .NET Core ? (Framework 4.6.25917.03), 64bit RyuJIT DEBUG
  Job-YAJUUM : .NET Core 2.1.0-preview1-25919-02 (Framework 4.6.25917.03), 64bit RyuJIT

MaxRelativeError=0.2  RemoveOutliers=True  Runtime=Core  
Toolchain=Core 2.1.0-preview1-25919-02  InvocationCount=6  LaunchCount=1  
TargetCount=2  UnrollFactor=1  WarmupCount=2  

```
|    Method |    Mean | Error |  StdDev |  Median |     Min |     Max |       Gen 0 |      Gen 1 |    Gen 2 | Allocated |
|---------- |--------:|------:|--------:|--------:|--------:|--------:|------------:|-----------:|---------:|----------:|
| serialize | 5.953 s |    NA | 5.422 s | 5.953 s | 2.120 s | 9.787 s | 206833.3333 | 68583.3333 | 500.0000 |  41.77 KB |

``` ini

BenchmarkDotNet=v0.10.10, OS=Windows 10 Redstone 2 [1703, Creators Update] (10.0.15063.726)
Processor=Intel Xeon CPU E5-1650 v4 3.60GHz, ProcessorCount=12
Frequency=3507499 Hz, Resolution=285.1034 ns, Timer=TSC
.NET Core SDK=2.1.1-preview-007118
  [Host]     : .NET Core ? (Framework 4.6.25917.03), 64bit RyuJIT DEBUG
  Job-HBVTON : .NET Core 2.1.0-preview1-25919-02 (Framework 4.6.25917.03), 64bit RyuJIT

MaxRelativeError=0.2  RemoveOutliers=True  Runtime=Core  
Toolchain=Core 2.1.0-preview1-25919-02  InvocationCount=6  LaunchCount=1  
TargetCount=2  UnrollFactor=1  WarmupCount=2  

```
|    Method |    Mean | Error |   StdDev |  Median |     Min |     Max |       Gen 0 |      Gen 1 |    Gen 2 | Allocated |
|---------- |--------:|------:|---------:|--------:|--------:|--------:|------------:|-----------:|---------:|----------:|
| serialize | 5.372 s |    NA | 0.6869 s | 5.372 s | 4.887 s | 5.858 s | 228833.3333 | 65833.3333 | 250.0000 |  54.79 KB |

ViktorHofer commented:
> 6 Cores with Locks

You mean threads?

Anipik commented:
Yah 6 threads

Anipik commented:
The code is available at 
https://github.com/Anipik/Benchmarking

stephentoub commented:
I don't understand what that test is doing.  The `serialize()` code is written as:
```C#
        [Benchmark]
        public void serialize()
        {
            int num = 8;
            //Console.WriteLine(num);
            for (int i = 0; i < num; i++)
            {

                StartWork += () =>
                {
                    Thread t = new Thread(() => serialization());
                    t.Start();
                };
            }
            StartWork();  
        }

        public delegate void StartWorkHandler();

        public event StartWorkHandler StartWork;
```
so each call to `serialize()` is going to create 8 delegates registered with an event handler and then invoke that event.  Those delegates aren't waiting for the thread to complete, though, so that's only then measuring how long it takes to start the threads, not actually do the work.  And each time the Benchmark is invoked on the same instance, which presumably is multiple times, it's going to register another 8 delegates (they're never removed), which would suggest that the first call to serialize spins up 8 threads, the next spins up 16, the next spins up 24, and so on.  And if it were measuring the right thing, it's also paying the overhead associated with creating and starting all those threads, which will add noise to the measurements.

I think the change is good.  I believe it could only help, and I think it's possible there are situations where it could help.  But I don't think that benchmark shows that.

ViktorHofer commented:
+1 what Stephen said. I looked at it earlier and it was clear to me that your implementation can't work the way you want it.

1. Consider creating Tasks instead of threads and more of them (> 50)
2. Create them not in the benchmark method but in the ctor of the benchmark class
3. Start and wait for their finish in the benchmark method: `await Tasks.WhenAll(...)`;

That removes the task creation costs from the benchmark measurement, what Stephen described.

Anipik commented:
@stephentoub @ViktorHofer  can you take a look at the code ..if its alright then I can start reviewing the performance
https://raw.githubusercontent.com/Anipik/Benchmarking/master/Program.cs

ViktorHofer commented:
I don't think you need all the types we test in the binary formatter. I suggest keeping 50-100 (primitive & reference types) for the tests.

ViktorHofer commented:
and of course you also don't need the blobs.

danmosemsft commented:
@Anipik move the setting of Culture out to around the body of `public void serialization()` so it doesn't happen 300 times (and get timed 300 times)

ViktorHofer commented:
you can completely get rid of the culture thing as it isn't relevant here.

danmosemsft commented:
You can remove the blobs with `, new string\[\] { "AAEAA.*" }`

ViktorHofer commented:
consider making the MemoryStream and BinaryFormatter static. As you are only serializing you can write into the same stream for all the types. Make sure to clean the memorystream in the IterationCleanup jmethod.

Anipik commented:
@ViktorHofer  @danmosemsft  I have addressed all of the feedback except memorystream . I tried clearing the memorystream with all sought of thing (flushing, seeking, disposing) but was not able to do so. Open to new ideas for reusing memory stream.
https://raw.githubusercontent.com/Anipik/Benchmarking/master/Program.cs

ViktorHofer commented:
```
[IterationCleanup]
        public void GlobalCleanup()
        {
            tasks.Clear();
            s = new MemoryStream();
            f = new BinaryFormatter();
        }
```

ViktorHofer commented:
try it with the above code. not sure if that helps.

danmosemsft commented:
Ideally you would randomize the array of object types but since you have many more tasks than likely threads/cores it should come out in the wash.

Anipik commented:
@ViktorHofer it worked.

Anipik commented:
``` ini

BenchmarkDotNet=v0.10.10, OS=Windows 10 Redstone 2 [1703, Creators Update] (10.0.15063.726)
Processor=Intel Xeon CPU E5-1650 v4 3.60GHz, ProcessorCount=12
Frequency=3507499 Hz, Resolution=285.1034 ns, Timer=TSC
.NET Core SDK=2.1.1-preview-007118
  [Host]     : .NET Core ? (Framework 4.6.25917.03), 64bit RyuJIT DEBUG
  Job-VGJVVY : .NET Core 2.1.0-preview1-25919-02 (Framework 4.6.25917.03), 64bit RyuJIT

MaxRelativeError=0.01  RemoveOutliers=True  Runtime=Core  
Toolchain=Core 2.1.0-preview1-25919-02  InvocationCount=1  LaunchCount=1  
TargetCount=10  UnrollFactor=1  WarmupCount=1  

```
|    Method |     Mean |    Error |   StdDev |   Median |      Min |      Max |       Gen 0 |      Gen 1 |      Gen 2 | Allocated |
|---------- |---------:|---------:|---------:|---------:|---------:|---------:|------------:|-----------:|-----------:|----------:|
| serialize | 560.1 ms | 17.69 ms | 11.70 ms | 556.3 ms | 548.8 ms | 581.4 ms | 343700.0000 | 69200.0000 | 41000.0000 | 133.27 KB |
``` ini

BenchmarkDotNet=v0.10.10, OS=Windows 10 Redstone 2 [1703, Creators Update] (10.0.15063.726)
Processor=Intel Xeon CPU E5-1650 v4 3.60GHz, ProcessorCount=12
Frequency=3507499 Hz, Resolution=285.1034 ns, Timer=TSC
.NET Core SDK=2.1.1-preview-007118
  [Host]     : .NET Core ? (Framework 4.6.25917.03), 64bit RyuJIT DEBUG
  Job-AYGDSO : .NET Core 2.1.0-preview1-25919-02 (Framework 4.6.25917.03), 64bit RyuJIT

MaxRelativeError=0.01  RemoveOutliers=True  Runtime=Core  
Toolchain=Core 2.1.0-preview1-25919-02  InvocationCount=1  LaunchCount=1  
TargetCount=10  UnrollFactor=1  WarmupCount=1  

```
|    Method |    Mean |    Error |   StdDev |  Median |     Min |     Max |       Gen 0 |       Gen 1 |      Gen 2 | Allocated |
|---------- |--------:|---------:|---------:|--------:|--------:|--------:|------------:|------------:|-----------:|----------:|
| serialize | 1.156 s | 0.0173 s | 0.0115 s | 1.153 s | 1.141 s | 1.173 s | 687400.0000 | 140300.0000 | 83000.0000 | 266.09 KB |

``` ini

BenchmarkDotNet=v0.10.10, OS=Windows 10 Redstone 2 [1703, Creators Update] (10.0.15063.726)
Processor=Intel Xeon CPU E5-1650 v4 3.60GHz, ProcessorCount=12
Frequency=3507499 Hz, Resolution=285.1034 ns, Timer=TSC
.NET Core SDK=2.1.1-preview-007118
  [Host]     : .NET Core ? (Framework 4.6.25917.03), 64bit RyuJIT DEBUG
  Job-WYOPMF : .NET Core 2.1.0-preview1-25919-02 (Framework 4.6.25917.03), 64bit RyuJIT

MaxRelativeError=0.01  RemoveOutliers=True  Runtime=Core  
Toolchain=Core 2.1.0-preview1-25919-02  InvocationCount=1  LaunchCount=1  
TargetCount=10  UnrollFactor=1  WarmupCount=1  

```
|    Method |    Mean |    Error |   StdDev |  Median |     Min |     Max |        Gen 0 |       Gen 1 |       Gen 2 | Allocated |
|---------- |--------:|---------:|---------:|--------:|--------:|--------:|-------------:|------------:|------------:|----------:|
| serialize | 2.483 s | 0.3149 s | 0.2083 s | 2.361 s | 2.334 s | 2.812 s | 1379700.0000 | 277500.0000 | 166000.0000 | 531.71 KB |

Anipik commented:
for 2000 and 4000 tasks using concurrent dictionary
``` ini

BenchmarkDotNet=v0.10.10, OS=Windows 10 Redstone 2 [1703, Creators Update] (10.0.15063.726)
Processor=Intel Xeon CPU E5-1650 v4 3.60GHz, ProcessorCount=12
Frequency=3507499 Hz, Resolution=285.1034 ns, Timer=TSC
.NET Core SDK=2.1.1-preview-007118
  [Host]     : .NET Core ? (Framework 4.6.25917.03), 64bit RyuJIT DEBUG
  Job-KAPCJY : .NET Core 2.1.0-preview1-25919-02 (Framework 4.6.25917.03), 64bit RyuJIT

MaxRelativeError=0.01  RemoveOutliers=True  Runtime=Core  
Toolchain=Core 2.1.0-preview1-25919-02  InvocationCount=1  LaunchCount=1  
TargetCount=10  UnrollFactor=1  WarmupCount=1  

```
|    Method |    Mean |    Error |   StdDev |  Median |     Min |     Max |       Gen 0 |       Gen 1 |      Gen 2 | Allocated |
|---------- |--------:|---------:|---------:|--------:|--------:|--------:|------------:|------------:|-----------:|----------:|
| serialize | 1.351 s | 0.0345 s | 0.0205 s | 1.355 s | 1.306 s | 1.374 s | 690000.0000 | 141300.0000 | 83000.0000 | 266.09 KB |

``` ini

BenchmarkDotNet=v0.10.10, OS=Windows 10 Redstone 2 [1703, Creators Update] (10.0.15063.726)
Processor=Intel Xeon CPU E5-1650 v4 3.60GHz, ProcessorCount=12
Frequency=3507499 Hz, Resolution=285.1034 ns, Timer=TSC
.NET Core SDK=2.1.1-preview-007118
  [Host]     : .NET Core ? (Framework 4.6.25917.03), 64bit RyuJIT DEBUG
  Job-MZVELF : .NET Core 2.1.0-preview1-25919-02 (Framework 4.6.25917.03), 64bit RyuJIT

MaxRelativeError=0.01  RemoveOutliers=True  Runtime=Core  
Toolchain=Core 2.1.0-preview1-25919-02  InvocationCount=1  LaunchCount=1  
TargetCount=10  UnrollFactor=1  WarmupCount=1  

```
|    Method |    Mean |    Error |   StdDev |  Median |     Min |     Max |        Gen 0 |       Gen 1 |       Gen 2 | Allocated |
|---------- |--------:|---------:|---------:|--------:|--------:|--------:|-------------:|------------:|------------:|----------:|
| serialize | 2.274 s | 0.0211 s | 0.0139 s | 2.273 s | 2.256 s | 2.301 s | 1379100.0000 | 278900.0000 | 166000.0000 | 531.71 KB |

Anipik commented:
The results are in release mode

danmosemsft commented:
What conclusions do you draw? I would also try with only a few specific types being serialized rather than a lot since presumably that will mean fewer writes to the dictionary.

Anipik commented:
Today morning run for Lock Version for 4000 tasks where as for concurrent dictionary version it has rose 
to 2.6 sec
I have updated the final code
``` ini

BenchmarkDotNet=v0.10.10, OS=Windows 10 Redstone 2 [1703, Creators Update] (10.0.15063.726)
Processor=Intel Xeon CPU E5-1650 v4 3.60GHz, ProcessorCount=12
Frequency=3507504 Hz, Resolution=285.1030 ns, Timer=TSC
.NET Core SDK=2.1.1-preview-007118
  [Host]     : .NET Core ? (Framework 4.6.25917.03), 64bit RyuJIT DEBUG
  Job-LGNGRT : .NET Core 2.1.0-preview1-25919-02 (Framework 4.6.25917.03), 64bit RyuJIT

MaxRelativeError=0.01  RemoveOutliers=True  Runtime=Core  
Toolchain=Core 2.1.0-preview1-25919-02  InvocationCount=1  LaunchCount=1  
TargetCount=10  UnrollFactor=1  WarmupCount=1  

```
|    Method |    Mean |    Error |   StdDev |  Median |     Min |     Max |        Gen 0 |       Gen 1 |       Gen 2 | Allocated |
|---------- |--------:|---------:|---------:|--------:|--------:|--------:|-------------:|------------:|------------:|----------:|
| serialize | 2.271 s | 0.0158 s | 0.0094 s | 2.269 s | 2.258 s | 2.284 s | 1382100.0000 | 280700.0000 | 166000.0000 | 531.71 KB |

Anipik commented:
Observations 

1.  For lower number of tasks around 1000. Locks perform better
2.  For # tasks around 2000. Both versions perform almost same. (These are after a series of tests)

For #tasks greater than 2000 , the trends are not conclusive and needs a bit more testing

danmosemsft commented:
Did you try cutting down to say 20 different types?

Anipik commented:
i am just wrapping up the trends for higher number of tasks. Then i will move on to types

ViktorHofer commented:
Make sure the invocation time takes at least 100ms so that the setup and cleanup costs won't spoil the result: https://github.com/dotnet/BenchmarkDotNet/blob/master/docs/guide/Advanced/SetupAndCleanup.md

Anipik commented:
6000, 8000 tasks with locks and dictionary
``` ini

BenchmarkDotNet=v0.10.10, OS=Windows 10 Redstone 2 [1703, Creators Update] (10.0.15063.726)
Processor=Intel Xeon CPU E5-1650 v4 3.60GHz, ProcessorCount=12
Frequency=3507504 Hz, Resolution=285.1030 ns, Timer=TSC
.NET Core SDK=2.1.1-preview-007118
  [Host]     : .NET Core ? (Framework 4.6.25917.03), 64bit RyuJIT DEBUG
  Job-ZOLPGT : .NET Core 2.1.0-preview1-25919-02 (Framework 4.6.25917.03), 64bit RyuJIT

MaxRelativeError=0.01  RemoveOutliers=True  Runtime=Core  
Toolchain=Core 2.1.0-preview1-25919-02  InvocationCount=1  LaunchCount=1  
TargetCount=10  UnrollFactor=1  WarmupCount=1  

```
|    Method |    Mean |    Error |   StdDev |  Median |     Min |     Max |        Gen 0 |       Gen 1 |       Gen 2 | Allocated |
|---------- |--------:|---------:|---------:|--------:|--------:|--------:|-------------:|------------:|------------:|----------:|
| serialize | 3.484 s | 0.0359 s | 0.0237 s | 3.482 s | 3.455 s | 3.526 s | 2074100.0000 | 421500.0000 | 249000.0000 | 797.34 KB |

``` ini

BenchmarkDotNet=v0.10.10, OS=Windows 10 Redstone 2 [1703, Creators Update] (10.0.15063.726)
Processor=Intel Xeon CPU E5-1650 v4 3.60GHz, ProcessorCount=12
Frequency=3507504 Hz, Resolution=285.1030 ns, Timer=TSC
.NET Core SDK=2.1.1-preview-007118
  [Host]     : .NET Core ? (Framework 4.6.25917.03), 64bit RyuJIT DEBUG
  Job-BXNRTG : .NET Core 2.1.0-preview1-25919-02 (Framework 4.6.25917.03), 64bit RyuJIT

MaxRelativeError=0.01  RemoveOutliers=True  Runtime=Core  
Toolchain=Core 2.1.0-preview1-25919-02  InvocationCount=1  LaunchCount=1  
TargetCount=10  UnrollFactor=1  WarmupCount=1  

```
|    Method |    Mean |    Error |   StdDev |  Median |     Min |     Max |        Gen 0 |       Gen 1 |       Gen 2 | Allocated |
|---------- |--------:|---------:|---------:|--------:|--------:|--------:|-------------:|------------:|------------:|----------:|
| serialize | 4.598 s | 0.0324 s | 0.0170 s | 4.596 s | 4.578 s | 4.628 s | 2761000.0000 | 567800.0000 | 333000.0000 |   1.04 MB |


``` ini

BenchmarkDotNet=v0.10.10, OS=Windows 10 Redstone 2 [1703, Creators Update] (10.0.15063.726)
Processor=Intel Xeon CPU E5-1650 v4 3.60GHz, ProcessorCount=12
Frequency=3507504 Hz, Resolution=285.1030 ns, Timer=TSC
.NET Core SDK=2.1.1-preview-007118
  [Host]     : .NET Core ? (Framework 4.6.25917.03), 64bit RyuJIT DEBUG
  Job-PWHSAY : .NET Core 2.1.0-preview1-25919-02 (Framework 4.6.25917.03), 64bit RyuJIT

MaxRelativeError=0.01  RemoveOutliers=True  Runtime=Core  
Toolchain=Core 2.1.0-preview1-25919-02  InvocationCount=1  LaunchCount=1  
TargetCount=10  UnrollFactor=1  WarmupCount=1  

```
|    Method |    Mean |    Error |   StdDev |  Median |     Min |     Max |        Gen 0 |       Gen 1 |       Gen 2 | Allocated |
|---------- |--------:|---------:|---------:|--------:|--------:|--------:|-------------:|------------:|------------:|----------:|
| serialize | 3.518 s | 0.0133 s | 0.0079 s | 3.519 s | 3.506 s | 3.532 s | 2071800.0000 | 421200.0000 | 249000.0000 | 797.34 KB |

``` ini

BenchmarkDotNet=v0.10.10, OS=Windows 10 Redstone 2 [1703, Creators Update] (10.0.15063.726)
Processor=Intel Xeon CPU E5-1650 v4 3.60GHz, ProcessorCount=12
Frequency=3507504 Hz, Resolution=285.1030 ns, Timer=TSC
.NET Core SDK=2.1.1-preview-007118
  [Host]     : .NET Core ? (Framework 4.6.25917.03), 64bit RyuJIT DEBUG
  Job-MQZOTU : .NET Core 2.1.0-preview1-25919-02 (Framework 4.6.25917.03), 64bit RyuJIT

MaxRelativeError=0.01  RemoveOutliers=True  Runtime=Core  
Toolchain=Core 2.1.0-preview1-25919-02  InvocationCount=1  LaunchCount=1  
TargetCount=10  UnrollFactor=1  WarmupCount=1  

```
|    Method |    Mean |    Error |   StdDev |  Median |     Min |     Max |        Gen 0 |       Gen 1 |       Gen 2 | Allocated |
|---------- |--------:|---------:|---------:|--------:|--------:|--------:|-------------:|------------:|------------:|----------:|
| serialize | 4.669 s | 0.0300 s | 0.0198 s | 4.664 s | 4.644 s | 4.706 s | 2766000.0000 | 558400.0000 | 333000.0000 |   1.04 MB |

Anipik commented:
``` ini

BenchmarkDotNet=v0.10.10, OS=Windows 10 Redstone 2 [1703, Creators Update] (10.0.15063.726)
Processor=Intel Xeon CPU E5-1650 v4 3.60GHz, ProcessorCount=12
Frequency=3507504 Hz, Resolution=285.1030 ns, Timer=TSC
.NET Core SDK=2.1.1-preview-007118
  [Host]     : .NET Core ? (Framework 4.6.25917.03), 64bit RyuJIT DEBUG
  Job-EJHZLN : .NET Core 2.1.0-preview1-25919-02 (Framework 4.6.25917.03), 64bit RyuJIT

RemoveOutliers=True  Runtime=Core  Toolchain=Core 2.1.0-preview1-25919-02  
InvocationCount=1  LaunchCount=1  TargetCount=10  
UnrollFactor=1  WarmupCount=1  

```
|    Method |    Mean |    Error |   StdDev |  Median |     Min |     Max |       Gen 0 |       Gen 1 |       Gen 2 | Allocated |
|---------- |--------:|---------:|---------:|--------:|--------:|--------:|------------:|------------:|------------:|----------:|
| serialize | 1.412 s | 0.0308 s | 0.0204 s | 1.403 s | 1.385 s | 1.448 s | 744100.0000 | 113500.0000 | 110800.0000 | 133.27 KB |

``` ini

BenchmarkDotNet=v0.10.10, OS=Windows 10 Redstone 2 [1703, Creators Update] (10.0.15063.726)
Processor=Intel Xeon CPU E5-1650 v4 3.60GHz, ProcessorCount=12
Frequency=3507504 Hz, Resolution=285.1030 ns, Timer=TSC
.NET Core SDK=2.1.1-preview-007118
  [Host]     : .NET Core ? (Framework 4.6.25917.03), 64bit RyuJIT DEBUG
  Job-WVIUUH : .NET Core 2.1.0-preview1-25919-02 (Framework 4.6.25917.03), 64bit RyuJIT

RemoveOutliers=True  Runtime=Core  Toolchain=Core 2.1.0-preview1-25919-02  
InvocationCount=1  LaunchCount=1  TargetCount=10  
UnrollFactor=1  WarmupCount=1  

```
|    Method |    Mean |    Error |   StdDev |  Median |     Min |     Max |        Gen 0 |       Gen 1 |       Gen 2 | Allocated |
|---------- |--------:|---------:|---------:|--------:|--------:|--------:|-------------:|------------:|------------:|----------:|
| serialize | 2.955 s | 0.2659 s | 0.1759 s | 2.836 s | 2.805 s | 3.185 s | 1496100.0000 | 229600.0000 | 224900.0000 | 266.09 KB |

BenchmarkDotNet=v0.10.10, OS=Windows 10 Redstone 2 [1703, Creators Update] (10.0.15063.726)
Processor=Intel Xeon CPU E5-1650 v4 3.60GHz, ProcessorCount=12
Frequency=3507504 Hz, Resolution=285.1030 ns, Timer=TSC
.NET Core SDK=2.1.1-preview-007118
  [Host]     : .NET Core ? (Framework 4.6.25917.03), 64bit RyuJIT DEBUG
  Job-OSDYYD : .NET Core 2.1.0-preview1-25919-02 (Framework 4.6.25917.03), 64bit RyuJIT

RemoveOutliers=True  Runtime=Core  Toolchain=Core 2.1.0-preview1-25919-02  
InvocationCount=1  LaunchCount=1  TargetCount=10  
UnrollFactor=1  WarmupCount=1  

```
|    Method |    Mean |    Error |   StdDev |  Median |     Min |     Max |        Gen 0 |       Gen 1 |       Gen 2 | Allocated |
|---------- |--------:|---------:|---------:|--------:|--------:|--------:|-------------:|------------:|------------:|----------:|
| serialize | 5.744 s | 0.0310 s | 0.0162 s | 5.742 s | 5.724 s | 5.780 s | 3018900.0000 | 464800.0000 | 443100.0000 | 531.71 KB |


``` ini

BenchmarkDotNet=v0.10.10, OS=Windows 10 Redstone 2 [1703, Creators Update] (10.0.15063.726)
Processor=Intel Xeon CPU E5-1650 v4 3.60GHz, ProcessorCount=12
Frequency=3507504 Hz, Resolution=285.1030 ns, Timer=TSC
.NET Core SDK=2.1.1-preview-007118
  [Host]     : .NET Core ? (Framework 4.6.25917.03), 64bit RyuJIT DEBUG
  Job-ZZBCJG : .NET Core 2.1.0-preview1-25919-02 (Framework 4.6.25917.03), 64bit RyuJIT

RemoveOutliers=True  Runtime=Core  Toolchain=Core 2.1.0-preview1-25919-02  
InvocationCount=1  LaunchCount=1  TargetCount=10  
UnrollFactor=1  WarmupCount=1  

```
|    Method |    Mean |    Error |   StdDev |  Median |     Min |     Max |       Gen 0 |       Gen 1 |       Gen 2 | Allocated |
|---------- |--------:|---------:|---------:|--------:|--------:|--------:|------------:|------------:|------------:|----------:|
| serialize | 1.574 s | 0.0835 s | 0.0553 s | 1.563 s | 1.509 s | 1.701 s | 744300.0000 | 113000.0000 | 110600.0000 | 133.27 KB |

``` ini

BenchmarkDotNet=v0.10.10, OS=Windows 10 Redstone 2 [1703, Creators Update] (10.0.15063.726)
Processor=Intel Xeon CPU E5-1650 v4 3.60GHz, ProcessorCount=12
Frequency=3507504 Hz, Resolution=285.1030 ns, Timer=TSC
.NET Core SDK=2.1.1-preview-007118
  [Host]     : .NET Core ? (Framework 4.6.25917.03), 64bit RyuJIT DEBUG
  Job-AGSCYW : .NET Core 2.1.0-preview1-25919-02 (Framework 4.6.25917.03), 64bit RyuJIT

RemoveOutliers=True  Runtime=Core  Toolchain=Core 2.1.0-preview1-25919-02  
InvocationCount=1  LaunchCount=1  TargetCount=10  
UnrollFactor=1  WarmupCount=1  

```
|    Method |    Mean |    Error |   StdDev |  Median |     Min |     Max |        Gen 0 |       Gen 1 |       Gen 2 | Allocated |
|---------- |--------:|---------:|---------:|--------:|--------:|--------:|-------------:|------------:|------------:|----------:|
| serialize | 2.864 s | 0.0226 s | 0.0134 s | 2.867 s | 2.842 s | 2.883 s | 1495800.0000 | 229400.0000 | 224300.0000 | 266.09 KB |


``` ini

BenchmarkDotNet=v0.10.10, OS=Windows 10 Redstone 2 [1703, Creators Update] (10.0.15063.726)
Processor=Intel Xeon CPU E5-1650 v4 3.60GHz, ProcessorCount=12
Frequency=3507504 Hz, Resolution=285.1030 ns, Timer=TSC
.NET Core SDK=2.1.1-preview-007118
  [Host]     : .NET Core ? (Framework 4.6.25917.03), 64bit RyuJIT DEBUG
  Job-NWRBKR : .NET Core 2.1.0-preview1-25919-02 (Framework 4.6.25917.03), 64bit RyuJIT

RemoveOutliers=True  Runtime=Core  Toolchain=Core 2.1.0-preview1-25919-02  
InvocationCount=1  LaunchCount=1  TargetCount=10  
UnrollFactor=1  WarmupCount=1  

```
|    Method |    Mean |    Error |   StdDev |  Median |     Min |     Max |        Gen 0 |       Gen 1 |       Gen 2 | Allocated |
|---------- |--------:|---------:|---------:|--------:|--------:|--------:|-------------:|------------:|------------:|----------:|
| serialize | 6.049 s | 0.4491 s | 0.2971 s | 5.856 s | 5.762 s | 6.460 s | 3012600.0000 | 465400.0000 | 442400.0000 | 531.71 KB |

Anipik commented:
the performance with dictionary improves as the number of dictionary access increases and writes decreases.
The trends does not vary much with increasing tasks.
In the second case in above table (2.955s) the average was about 2.8 sec only. I performed 3 to 5 trials . 2.935 is the result of the last trial

Anipik commented:
The iteration time is greater than 500ms for all the runs

danmosemsft commented:
@Anipik It is not clear what table corresponds to what scenario/code. You need to paste in the processor count, etc - all the benchmark.NET metadata (at least not more than once). 

Please reformat all your reliable results into a single table, with added left column(s): a column for whether it is "before" or "after" your changes, and also column(s) describing the scenario parameters (eg number of types, number of tasks).

Anipik commented:
```
|    Tasks  |   Types |   Mean   |  Median  | Before/After |
|---------- |--------:|---------:|---------:|------------: |
|   1000    |   23    | 1.412    | 1.403 s  | After        |
|---------- |--------:|---------:|---------:|------------: |
| 1000      |   23    |  1.574   | 1.509 s  | Before       | 
|---------- |--------:|---------:|---------:|------------: |
| 1000      |   23    | 2.955    | 2.836 s  | After        | 
|---------- |--------:|---------:|---------:|------------: |
| 1000      |   23    | 2.864    | 2.867 s  | Before       | 
|---------- |--------:|---------:|---------:|------------: |
| 1000      |   23    | 5.742    | 5.724 s  | After        | 
|---------- |--------:|---------:|---------:|------------: |
| 1000      |   23    | 5.856    | 5.762 s  | Before       |

Anipik commented:
``` ini

RemoveOutliers=True  Runtime=Core  Toolchain=Core 2.1.0-preview1-25919-02  
InvocationCount=1  LaunchCount=1  TargetCount=10  
UnrollFactor=1  WarmupCount=1  

```
|    Tasks  |   Types |   Mean   |  Median  | Before/After |
|---------- |--------:|---------:|---------:|------------: |
| 1000      |   23    | 1.412    | 1.403 s  | After        |
|---------- |--------:|---------:|---------:|------------: |
| 1000      |   23    |  1.574   | 1.509 s  | Before       | 
|---------- |--------:|---------:|---------:|------------: |
| 2000      |   23    | 2.955    | 2.836 s  | After        | 
|---------- |--------:|---------:|---------:|------------: |
| 2000      |   23    | 2.864    | 2.867 s  | Before       | 
|---------- |--------:|---------:|---------:|------------: |
| 4000      |   23    | 5.742    | 5.724 s  | After        | 
|---------- |--------:|---------:|---------:|------------: |
| 4000      |   23    | 5.856    | 5.762 s  | Before       |

Anipik commented:
``` ini

RemoveOutliers=True  Runtime=Core  Toolchain=Core 2.1.0-preview1-25919-02  
InvocationCount=1  LaunchCount=1  TargetCount=10  
UnrollFactor=1  WarmupCount=1  

```
|    Tasks  |   Types |   Mean   |  Median  | Before/After |
|---------- |--------:|---------:|---------:|------------: |
| 1000      |   23    | 1.412    | 1.403 s  | After        |
| 1000      |   23    |  1.574   | 1.509 s  | Before       | 
| 2000      |   23    | 2.955    | 2.836 s  | After        | 
| 2000      |   23    | 2.864    | 2.867 s  | Before       | 
| 4000      |   23    | 5.742    | 5.724 s  | After        | 
| 4000      |   23    | 5.856    | 5.762 s  | Before       |

danmosemsft commented:
That's good enough to demonstrate an improvement. Thanks @Anipik .
